backend/src/apiserver/resource/resource_manager.go

```diff
@@ -2048,3 +2058,137 @@ func (r *ResourceManager) GetTask(taskId string) (*model.Task, error) {
 	}
 	return task, nil
 }
+
+// ensureSemaphoreConfigMap parses the pipeline manifest and creates semaphore ConfigMaps if needed
+func (r *ResourceManager) ensureSemaphoreConfigMap(ctx context.Context, namespace, manifest string) error {
+	// Parse the manifest to extract semaphore configuration
+	semaphoreKey, mutexName, err := r.extractSemaphoreConfig(manifest)
+	if err != nil {
+		glog.Warningf("Failed to parse semaphore configuration from manifest: %v", err)
+		return nil // Don't fail run creation for semaphore parsing issues
+	}
+
+	// Create ConfigMap for semaphore_key if present
+	if semaphoreKey != "" {
+		if err := r.semaphoreManager.EnsureSemaphoreConfigMap(ctx, namespace, semaphoreKey); err != nil {
+			glog.Errorf("Failed to ensure semaphore ConfigMap for key %s: %v", semaphoreKey, err)
+			return nil // Don't fail run creation for ConfigMap issues
+		}
+	}
+
+	// Create ConfigMap for mutex_name if present (same logic, treat as semaphore with limit 1)
+	if mutexName != "" {
```

Comment:
I think it'd be better to use Argo Workflows' official mutex support. For example:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: synchronization-wf-level-
  namespace: foo
spec:
  synchronization:
    mutexes:
      - name: bar
```